### DISCUSSION TITLE
Get Smooth & Straight Edges without curves and overlapping nodes

### DISCUSSION THREAD

**Discussion by vaibhav-systango** [xyflow/xyflow#2199] [Q&A] [answered]
Currently my react flow implementation looks like this:
<img width="1101" alt="Screen Shot 2022-06-06 at 1 52 41 PM" src="https://user-images.githubusercontent.com/25978575/172123871-38197ee4-1e43-460b-8b94-8eac8758ffa8.png">

I want it to look neat and clean like this:
https://dbdiagram.io/d/62554b9e2514c979031605c4

@moklick @chrtze

**Comment by bcakmakoglu** [comment, marked as answer]
You can set edge types by using the `edge.type` property, like `type: 'smoothstep'` or `type: 'step'` etc..

There's an example showing the different built-in types [here](https://reactflow.dev/docs/examples/edges/edge-types/).

If you want handles to be at different positions in your custom node, you should place them there using `<Handle position="right" type="source" />`.

As for the non-overlapping part: Built-in edges don't check for collision, that's something you'd have to implement yourself.

**Reply by vaibhav-systango** [reply]
alright also can you even tell me how can we have dynamic positioning of connectors, for example if a table is on the right it's connnector should come on the right of the current node, and if a table is on top its conection should come on top. Also can we have multi edges connection points instead of a single multi edge connection from a single point ?

**Reply by bcakmakoglu** [reply]
Do you mean Handles when you say "connectors"? The small dots that are used as the positions where edges connect, right?
You can place Handles dynamically, using `position={handlePosition}` - just make sure that you call `updateNodeInternals` afterwards, otherwise React Flow doesn't notice the change in handle positions.
You can also generate a dynamic number of Handles (I believe that's what you want, multiple connection points, right?). 
Something like `handles.map(handle => <Handle {...handle} />)`.
Also make sure to call `updateNodeInternals` afterwards here.
[Here's the docs for `updateNodeInternals`.](https://reactflow.dev/docs/api/hooks/use-update-node-internals/)

[Also check the Custom Node example, it uses two Handles on the right side.](https://reactflow.dev/docs/examples/nodes/custom-node/)

**Reply by tisoap** [reply]
> how can we have dynamic positioning of connectors, for example if a table is on the right it's connnector should come on the right of the current node, and if a table is on top its conection should come on top

@vaibhav-systango  React Flow has an example on how to implement [Floating Edges](https://reactflow.dev/docs/examples/edges/floating-edges/), which do something similar to what you described

> As for the non-overlapping part: Built-in edges don't check for collision, that's something you'd have to implement yourself

@bcakmakoglu @vaibhav-systango I've implemented a custom edge that does exactly that, you can check it out here:
[@tisoap/react-flow-smart-edge](https://www.npmjs.com/package/@tisoap/react-flow-smart-edge)

**Reply by vaibhav-systango** [reply]
Hey the non colliding edge thing worked for me from the smart step edge thanks a lot for your extension @tisoap 

Rest let me check the floating edge example for controlling the postion of connectors and get back 💯

**Reply by vaibhav-systango** [reply]
@tisoap @moklick any example where we can see both the react smart step along with this floating edges being used together ?

**Reply by tisoap** [reply]
Newest version of my library has a `SmartFloatingEdge` export, check the documentation!

**Comment by vaibhav-systango** [comment]
For the above example If I manually arrange it, the ideal view is shown below but I want to achieve this in an automated fashion:
<img width="1173" alt="Screen Shot 2022-06-06 at 2 11 57 PM" src="https://user-images.githubusercontent.com/25978575/172127138-dac64f65-74af-4935-b7a4-ac7075855c4a.png">

**Reply by moklick** [reply]
Please check out the examples. You can find this one in the [`dagre tree` example](https://reactflow.dev/docs/examples/layout/dagre/)

**Reply by vaibhav-systango** [reply]
Thanks a lot @moklick @bcakmakoglu for the help here, I'll check these. Also there are any other layouting examples that can be used with dagre so that node with max conections comes in the middle and then the one with lesser connections surrround it something similar to a star schema or snowflake schema views

**Comment by vaibhav-systango** [comment]
@moklick @bcakmakoglu @tisoap when I use smart step edge from here:
[@tisoap/react-flow-smart-edge](https://www.npmjs.com/package/@tisoap/react-flow-smart-edge)
the custom edge markers get misaligned when i try to change the node layouts why is this happening ? do we have any solutions around this.
please see the gif shared below:
![smartedge](https://user-images.githubusercontent.com/25978575/172391711-e0bf2541-33f8-4d06-8ea4-7910d43af135.gif)

**Reply by moklick** [reply]
We are not maintaining this package. Please open an issue in the smart edge repo.

**Comment by vaibhav-systango** [comment]
@moklick @bcakmakoglu floating edges doesn't seem to work with child nodes ? Any solution to it ?

**Reply by vaibhav-systango** [reply]
https://codesandbox.io/s/sandpack-project-forked-ev1563?file=/App.js
here's  a sample sandbox @moklick @bcakmakoglu @chrtze
